### PR TITLE
is_ipv4_compat and is_ipv4_mapped expression for coprocessor

### PR DESCRIPTION
**Pull request by yaalsn**
Signed-off-by: yaalsn <yaalsn@gmail.com>

## What have you changed? (mandatory)
Added is_ipv4_compat and is_ipv4_mapped expression for coprocessor

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:
- Summarize your change (mandatory)

1. modify built_miscellaneous.rs by adding two functions which are named is_ipv4_compat and is_ipv4_mapped.
2. modify scalar_function.rs to apply above changes.
3. add two constants: IPV6_LENGTH and IPV4_LENGTH, and override original number with the constant in built_miscellaneous.rs

- How does this PR work? Need a brief introduction for the changed logic (optional)
- Separately describe each logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

Please **NOTE** that:
- Do not assume reviewers understand the original issue

## What are the type of the changes? (mandatory)
- New feature (change which adds functionality)

## How has this PR been tested? (mandatory)
unit test

## Does this PR affect documentation (docs) update? (mandatory)
no

## Does this PR affect tidb-ansible update? (mandatory)
no

## Refer to a related PR or issue link (optional)
#3275

## Add a few positive/negative examples (optional)

### CODE REVIEW

**Comment by yaalsn** [comment]
emmm...something wrong with jekins?
```shell
uname -a  — Shell Script <1s
Task okhttp3.RealCall$AsyncCall@203993e6 rejected from java.util.concurrent.ThreadPoolExecutor@47cbb43e[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 138]
```

**Comment by yaalsn** [comment]
PTAL @siddontang

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs [hunk 1]
```diff
@@ -28,6 +28,48 @@ impl ScalarFunc {
...
     }
 
+    pub fn is_ipv4_compat<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<i64>> {
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        //Not an IPv6 address, return 0
+        if input.len() != 16 {
```
Comment (rleungx) [inline src/coprocessor/dag/expr/builtin_miscellaneous.rs]: How about using a constant here?

**Comment by rleungx** [comment]
@AndreMouche @crazycs520 PTAL.

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs (hunk 1 above)
Comment (yaalsn) [reply src/coprocessor/dag/expr/builtin_miscellaneous.rs]: I have not find a constant about ipv6 length in rust, if you know please tell me...and It seems there are two ways if using a constant, one is declaring it in this file or somewhere else, another is calculate the struct in6_addr size, which one is better?

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs [hunk 2]
```diff
@@ -28,6 +28,48 @@ impl ScalarFunc {
...
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        // Not an IPv6 address, return 0
+        if input.len() != 16 {
+            return Ok(Some(0));
+        }
+
+        let input_bytes: &[u8; 16] = input.as_ref().try_into().unwrap();
+        let prefix_compat = vec![
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        ];
```
Comment (kennytm) [inline src/coprocessor/dag/expr/builtin_miscellaneous.rs]: As I mentioned in https://github.com/tikv/tikv/pull/4405#pullrequestreview-216529973 you don't need to create a Vec.

Comment (yaalsn) [reply src/coprocessor/dag/expr/builtin_miscellaneous.rs]: thx for review, fixed

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs (hunk 1 above)
Comment (rleungx) [reply src/coprocessor/dag/expr/builtin_miscellaneous.rs]: I think declaring it in this file is ok.

Comment (yaalsn) [reply src/coprocessor/dag/expr/builtin_miscellaneous.rs]: I add two constants: IPV6_LENGTH and IPV4_LENGTH, and override original number with the constant in built_miscellaneous.rs

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs [hunk 3]
```diff
@@ -28,6 +31,46 @@ impl ScalarFunc {
...
+        row: &'a [Datum],
+    ) -> Result<Option<i64>> {
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        // Not an IPv6 address, return 0
+        if input.len() != IPV6_LENGTH {
+            return Ok(Some(0));
+        }
+
+        let input_bytes: &[u8; 16] = input.as_ref().try_into().unwrap();
+        let prefix_compat: [u8; 12] = [0x00; 12];
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_miscellaneous.rs]: Could we define it a constant?

File: src/coprocessor/dag/expr/builtin_miscellaneous.rs [hunk 4]
```diff
@@ -28,6 +31,46 @@ impl ScalarFunc {
...
+        if !input_bytes.starts_with(&prefix_compat) {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_miscellaneous.rs]: Could we use 
```
input.starts_with(&prefix_compat)
```
directly?

Comment (yaalsn) [reply src/coprocessor/dag/expr/builtin_miscellaneous.rs]: Awesome!

**Comment by rleungx** [comment]
@yaalsn Thanks! Merged.

**Pull request merged**